wp_api/src/wordpress_org/client.rs

```diff
@@ -0,0 +1,249 @@
+        Self::parse(response)
+    }
+}
+
+impl WordPressOrgApiClient {
+    fn plugin_info_api_url() -> Url {
+        Url::parse("https://api.wordpress.org/plugins/info/1.2/").expect("The URL is valid")
+    }
+
+    fn parse<T>(response: WpNetworkResponse) -> Result<T>
+    where
+        T: DeserializeOwned,
+    {
+        match response.status_code {
+            200 => serde_json::from_slice(&response.body).map_err(|e| {
+                WordPressOrgApiClientError::ResponseParsingError {
+                    reason: format!("Failed to parse response body as JSON: {}", e),
+                    response: String::from_utf8_lossy(&response.body).to_string(),
+                }
+            }),
+            _ => Err(WordPressOrgApiClientError::UnexpectedStatusCodeError {
+                status_code: response.status_code,
+                response: String::from_utf8_lossy(&response.body).to_string(),
+            }),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, uniffi::Enum)]
+pub enum WordPressOrgApiPluginDirectoryCategory {
+    New,
+    Popular,
+    Updated,
+    TopRated,
+}
+
+impl WordPressOrgApiPluginDirectoryCategory {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            WordPressOrgApiPluginDirectoryCategory::New => "new",
+            WordPressOrgApiPluginDirectoryCategory::Popular => "popular",
+            WordPressOrgApiPluginDirectoryCategory::Updated => "updated",
+            WordPressOrgApiPluginDirectoryCategory::TopRated => "top-rated",
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, thiserror::Error, uniffi::Error)]
+pub enum WordPressOrgApiClientError {
+    #[error(
+        "Request execution failed!\nStatus Code: '{:?}'.\nResponse: '{}'",
+        status_code,
+        reason
+    )]
+    RequestExecutionFailed {
+        status_code: Option<u16>,
+        reason: String,
+    },
+    #[error("Error while parsing. \nReason: {}\nResponse: {}", reason, response)]
+    ResponseParsingError { reason: String, response: String },
+    #[error(
+        "Received a response with an unexpected status code. \nStatus code: {}\nResponse: {}",
+        status_code,
+        response
+    )]
+    UnexpectedStatusCodeError { status_code: u16, response: String },
+}
+
+impl From<RequestExecutionError> for WordPressOrgApiClientError {
+    fn from(e: RequestExecutionError) -> Self {
+        match e {
+            RequestExecutionError::RequestExecutionFailed {
+                status_code,
+                reason,
+            } => WordPressOrgApiClientError::RequestExecutionFailed {
+                status_code,
+                reason,
+            },
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use futures::lock::Mutex;
+
+    use super::*;
+    use crate::{
+        request::{endpoint::media_endpoint::MediaUploadRequest, RequestExecutor},
+        MediaUploadRequestExecutionError,
+    };
+    use std::sync::Arc;
+
+    #[derive(Debug)]
+    struct MockRequestExecutor {
```

Comment:
> In my opinion, dealing with the request executor in the unit test level is not necessary. The added complexity doesn't bring much to the table when all we really need to test is the url building.

The two tests I added are not so much about URL/request building, though. Those tests are about ensuring `fields=icons` is included in API requests, and the right pagination parameters are used (i.e., `page` instead of `request[page]`).

Also, I want to write tests that check application behaviors rather than their implementation details. By "application behaviors", I mean assertions like a GET request whose URL includes query `search=xxx` is sent out when the `search_plugin()` function is called. By "implementation details", I mean assertions like `PluginSearchRequest(search: "xxx").build_request().url.contains("search=xxx")`. I'm not saying testing implementation details is not important, though. It's just not what the tests in this PR are about.

> If the url building happens on its own, then it'd be easy to unit test that and I'd really prefer if we did that.

Sure. I can look into that.

Reply:
> I would also prefer if we don't add tokio as a dependency if we can avoid it.

Can you talk a bit more about this?

Reply:
> Also, I want to write tests that check application behaviors rather than their implementation details.

Wouldn't these be integration tests though?

---

I am not sure I understand the benefit of dealing with a mock executor when we can check the details of the `WpNetworkRequest` that will be executed. Although I don't think even _that_ is necessary because testing the `url` should be more than enough. Since it's a GET request, everything about page, fields, search etc will be part of the url query pairs.

---

> Can you talk a bit more about this?

We want to remain async executor agnostic within `wp_api`, as we don't want to ship an executor as part of the library. This was something @jkmassel was specifically worried about when we first worked on the prototype.

Adding `tokio` as a dev dependency does not break that, so it's not an issue. However, I feel like if it's there, it's just one step away from making it as a regular dependency - which when I say it out loud, I didn't find myself convincing 😅 I guess in this case, it's easy to not require it, so I guess I'd rather not - until we really need it.

Reply:
> Wouldn't these be integration tests though?

No, I would not consider them as integration tests. I thought it's pretty common to use test doubles to take dependencies (HTTP client in this particular case) out of the test env in unit tests.

Reply:
Although that's definitely common, it's generally because that's the only way to test something - mainly because of poor design either in the language/framework, or the implementation itself.

Otherwise, unit testing requires testing the smallest unit possible - hence the name _unit_ testing. What you were doing in the tests - note that I haven't re-reviewed, was use a mock executor just to get the url information out of it and then check the details of it. Doesn't that sound unnecessary when you could just test the url itself?

Testing the smallest possible unit not only makes it more useful, but it's also a lot easier to understand what the test is trying to achieve. For example, in the implementation, you had an assertion to check only one request was being made - and you also implemented the logic for tracking how many requests are made within a mock executor. If I didn't have any context about the code base, looking at that test I'd think that it's possible and worrisome that there might be multiple requests. That's not a possibility, but even if it was, it'd be the wrong place to test it, simply because we'd have to do it for every single endpoint we implement.